### PR TITLE
NR validation before file and pay

### PR DESCRIPTION
**Pull request by lekshmimallika-aot**
*Issue #:* /bcgov/entity#4037

*Description of changes:*
Added an error dialog
Modified file and pay method to validate NR for a named company
Added tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -279,12 +287,31 @@ export default class App extends Mixins(BcolMixin, DateMixin, FilingTemplateMixi
           this.saveErrorDialog = true
       }
     })
+
+    this.$root.$on('name-request-invalid-error', async error => {
+      console.log('NR error during File and Pay =', error)
```
Comment (severinbeauvais) [inline src/App.vue:292]: If this console.log message is needed then you'll have to disable the lint error.
Also below.

File: src/App.vue [hunk 2]
```diff
@@ -279,12 +287,31 @@ export default class App extends Mixins(BcolMixin, DateMixin, FilingTemplateMixi
...
     // stop listening for save error event
     this.$root.$off('save-error-event')
+    this.$root.$off('name-request-invalid-errort')
+    this.$root.$off('name-request-retrieve-error')
+  }
+
+  private goToManageBusinessDashboard () : void {
+    this.fileAndPayInvalidNameRequestDialog = false
+    const manageBusinessUrl = `${sessionStorage.getItem('AUTH_URL')}business`
+    window.onbeforeunload = null
```
Comment (severinbeauvais) [inline src/App.vue:313]: You should clear the 'haveChanges' flag instead.

File: src/App.vue [hunk 3]
```diff
@@ -207,6 +214,7 @@ export default class App extends Mixins(BcolMixin, DateMixin, FilingTemplateMixi
   private haveData: boolean = false
   private saveErrors: Array<object> = []
   private saveWarnings: Array<object> = []
+  private fileAndPayInvalidNameRequestDialog: boolean = false
```
Comment (severinbeauvais) [inline src/App.vue:217]: You may have to add this flag to isErrorDialog() and resetFlags().

File: src/components/common/Actions.vue [hunk 4]
```diff
@@ -207,6 +221,26 @@ export default class Actions extends Mixins(DateMixin, FilingTemplateMixin, Lega
     }
   }
 
+  /** Fetches NR and validates it. */
+  private async validateNameRequest (nrNumber: string): Promise<void> {
```
Comment (severinbeauvais) [inline src/components/common/Actions.vue:225]: I noticed this is similar to processNameRequest() in App.vue. Can we name both functions the same? (I like this function's name better.)

File: src/App.vue (hunk 2 above)
Comment (lekshmimallika-aot) [reply src/App.vue:313]: This is not for the custom popup. This is to disable the browsers default behavior

File: src/components/common/Actions.vue [hunk 5]
```diff
@@ -171,6 +176,15 @@ export default class Actions extends Mixins(DateMixin, FilingTemplateMixin, Lega
       window.scrollTo({ top: 1250, behavior: 'smooth' })
       return
     }
+    // If it is a named company IA, validate NR before filing submission
+    if (this.isNamedBusiness && this.getNameRequestNumber) {
```
Comment (severinbeauvais) [inline src/components/common/Actions.vue:180]: You don't need both of these. See how they're declared.

File: src/components/common/Actions.vue (hunk 4 above)
Comment (lekshmimallika-aot) [reply src/components/common/Actions.vue:225]: The other one has more things to it like setting the data to the store and all. This is just validating the NR

File: tests/unit/FileAndPayInvalidNameRequestDialog.spec.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    expect(wrapper.findAll('p').at(0).text())
+      .toContain('The Name Request NR 1234567 and the Incorporation Application for')
+    expect(wrapper.findAll('p').at(0).text())
+      .toContain('My Name Request Inc. are no longer valid.')
+    expect(wrapper.findAll('p').at(1).text())
+      .toContain('If you still wish to incorporate a Benefit Company, please contact Registry Staff ' +
+        'as soon as possible.')
+    expect(wrapper.findAll('p').at(2).text())
+      .toContain('Registries contact information:')
+    expect(wrapper.findAll('div > div').at(0).text()).toContain('IMPORTANT:')
```
Comment (severinbeauvais) [inline tests/unit/FileAndPayInvalidNameRequestDialog.spec.ts:35]: These selectors are fragile and suggest that a classname should be used.

File: src/components/common/Actions.vue (hunk 5 above)
Comment (lekshmimallika-aot) [reply src/components/common/Actions.vue:180]: True. Will change it

File: src/App.vue (hunk 1 above)
Comment (lekshmimallika-aot) [reply src/App.vue:292]: I am not sure whether I understood this. I dont see any lint error locally and in PR checks

File: src/App.vue (hunk 2 above)
Comment (severinbeauvais) [reply src/App.vue:313]: What default behaviour?

File: src/App.vue (hunk 1 above)
Comment (severinbeauvais) [reply src/App.vue:292]: In GH for this PR, go to Files Changed tab and look through the code. There are 2 lines with a GHA warning.

File: src/components/common/Actions.vue (hunk 4 above)
Comment (severinbeauvais) [reply src/components/common/Actions.vue:225]: I was hoping to reconcile the 2 functions that do something very similar, by naming them the same. (I realize they do slightly different things so they can be merged.) It just doesn't feel very DRY at this point.
How about adding a comment to each function saying it's similar to the other one?

File: src/App.vue (hunk 2 above)
Comment (lekshmimallika-aot) [reply src/App.vue:313]: Nvm. I got it

File: src/App.vue (hunk 3 above)
Comment (lekshmimallika-aot) [reply src/App.vue:217]: Done

File: src/App.vue (hunk 2 above)
Comment (lekshmimallika-aot) [reply src/App.vue]: Done

File: src/components/common/Actions.vue (hunk 5 above)
Comment (lekshmimallika-aot) [reply src/components/common/Actions.vue]: Done

File: tests/unit/FileAndPayInvalidNameRequestDialog.spec.ts (hunk 6 above)
Comment (lekshmimallika-aot) [reply tests/unit/FileAndPayInvalidNameRequestDialog.spec.ts]: Done

File: src/App.vue (hunk 1 above)
Comment (lekshmimallika-aot) [reply src/App.vue:293]: ok

**Pull request merged**